### PR TITLE
[type:feature] add rule cache

### PR DESCRIPTION
**Pull request by moremind**
about #4120

I will refactor it by trie when uri condition

Make sure that:

- [ ] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor-guide).
- [ ] You submit test cases (unit or integration tests) that back your changes.
- [ ] Your local test passed `./mvnw clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

File: shenyu-bootstrap/src/main/resources/application.yml [hunk 1]
```diff
@@ -68,7 +68,8 @@ management:
 
 shenyu:
   matchCache:
-    enabled: false
+    selectorEnabled: false
+    ruleEnabled: true
```
Comment (yu199195) [inline shenyu-bootstrap/src/main/resources/application.yml:72]: maybe add rule MaxFreeMemory ？

Comment (moremind) [reply shenyu-bootstrap/src/main/resources/application.yml:72]: maxFreeMemory means JVM can use available memory. if the memory summary is 512MB, and maxFreeMemory is 256MB, then the selector and rule cache can use 256MB.

Comment (yu199195) [reply shenyu-bootstrap/src/main/resources/application.yml:72]: I meant to set up JVM memory separately

Comment (moremind) [reply shenyu-bootstrap/src/main/resources/application.yml:72]: selector give JVM selectorMaxFreeMemory, rule gives too? then the final maxFreeMemory = selectorMaxFreeMemory + ruleMaxFreeMemory.

Comment (yu199195) [reply shenyu-bootstrap/src/main/resources/application.yml:72]: have 2 paramter， selector Memory and rule FreeMemory

Comment (moremind) [reply shenyu-bootstrap/src/main/resources/application.yml:72]: yes, I will change it

**Review by yu199195 — APPROVED** [review]

**Pull request merged**